aioesphomeapi/connection.py

```diff
@@ -305,19 +311,23 @@ async def _do_connect() -> None:
             if login:
                 await self.login()
 
-        try:
-            # Allow 2 minutes for connect; this is only as a last measure
-            # to protect from issues if some part of the connect process mistakenly
-            # does not have a timeout
-            async with async_timeout.timeout(120.0):
-                await _do_connect()
-        except Exception:  # pylint: disable=broad-except
-            # Always clean up the connection if an error occured during connect
-            self._connection_state = ConnectionState.CLOSED
-            await self._cleanup()
-            raise
-
-        self._connect_complete = True
+        # A connection lock must be created to avoid potential issues where
+        # connect has succeeded but not yet returned, followed by a disconnect.
+        # See esphome/aioesphomeapi#258 for more information
+        async with self._connect_lock:
```

Comment:
This makes sense but I think the disconnect needs to run under the lock as well

Reply:
This shouldn't be necessary since the disconnect isn't doing anything that should be conflicting with the connection process. Aka there should be no additional packets coming in to connect the device while it's disconnecting as was the case with connecting.